File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */
     function reverse(head) {
       
        let current = head;
        let prev = null;
        while (current != null) {
            let temp = current.next;
            current.next = prev;
            prev = current;
            current = temp;
        }
        return prev
    }
    function findlength(head) {
        let length = 0;
        let current = head;

        while (current != null) {
            length++;

            current = current.next
        }
        return length;

    }
var reverseKGroup = function (head, k) {



    if (k==1 || head==null || head.next==null){
        return head;
    }

    if (k>findlength(head)){
        return head;
    }

    let headtemp = head;
    let tailofLL= head;
    for(let i=0;i<k-1;i++){
        tailofLL= tailofLL.next;

    }
    let next = tailofLL.next;
    tailofLL.next = null;
    let newhead = reverse(headtemp);
    headtemp.next = reverseKGroup(next, k);
    return newhead



};

File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    int Length(ListNode* head) {
        int l = 0;
        while (head) {
            l++;
            head = head->next;
        }
        return l;
    }
    ListNode* reverse(ListNode* head) {
        ListNode* prev = NULL;
        ListNode* current = head;
        while (current) {
            ListNode* temp = current->next;
            current->next = prev;
            prev = current;
            current = temp;
        }
        return prev;
    }
    ListNode* reverseKGroup(ListNode* head, int k) {
        if(Length(head)<k){
            return head;
        }
        ListNode* newhead = head;
        for(int i=1;i<=k-1;i++){
            newhead= newhead->next;
        }
        ListNode* next = newhead;
        next = newhead->next;
        newhead->next = NULL;
        ListNode* rerversed = reverse(head);
        head->next = reverseKGroup(next,k);
        return newhead;
    }
};